flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/source/DynamicFilteringData.java

```diff
@@ -0,0 +1,207 @@
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+import static org.apache.flink.util.Preconditions.checkNotNull;
+
+/** Data for dynamic filtering. */
+public class DynamicFilteringData implements Serializable {
+
+    public static final Set<LogicalTypeRoot> SUPPORTED_TYPES;
+
+    static {
+        Set<LogicalTypeRoot> supported = new HashSet<>();
+        supported.add(LogicalTypeRoot.INTEGER);
+        supported.add(LogicalTypeRoot.BIGINT);
+        supported.add(LogicalTypeRoot.SMALLINT);
+        supported.add(LogicalTypeRoot.TINYINT);
+        supported.add(LogicalTypeRoot.VARCHAR);
+        supported.add(LogicalTypeRoot.CHAR);
+        supported.add(LogicalTypeRoot.DATE);
+        supported.add(LogicalTypeRoot.TIME_WITHOUT_TIME_ZONE);
+        supported.add(LogicalTypeRoot.TIMESTAMP_WITH_LOCAL_TIME_ZONE);
+        supported.add(LogicalTypeRoot.TIMESTAMP_WITHOUT_TIME_ZONE);
+        SUPPORTED_TYPES = Collections.unmodifiableSet(supported);
+    }
+
+    private final TypeInformation<RowData> typeInfo;
+    private final RowType rowType;
+
+    /** The list should be sorted and distinct. */
+    private final List<byte[]> serializedData;
+
+    /** Whether the data actually does filter. If false, everything is considered contained. */
+    private final boolean isFiltering;
+
+    private transient volatile boolean prepared = false;
+    private transient Map<Integer, List<RowData>> dataMap;
+    private transient RowData.FieldGetter[] fieldGetters;
+
+    public DynamicFilteringData(
+            TypeInformation<RowData> typeInfo,
+            RowType rowType,
+            List<byte[]> serializedData,
+            boolean isFiltering) {
+        this.typeInfo = checkNotNull(typeInfo);
+        this.rowType = checkNotNull(rowType);
+        this.serializedData = checkNotNull(serializedData);
+        this.isFiltering = isFiltering;
+    }
+
+    public boolean isFiltering() {
+        return isFiltering;
+    }
+
+    public RowType getRowType() {
+        return rowType;
+    }
+
+    public boolean contains(RowData row) {
+        if (!isFiltering) {
+            return true;
+        } else if (row.getArity() != rowType.getFieldCount()) {
+            throw new TableException("The arity of RowData is different");
+        } else {
+            prepare();
+            List<RowData> mayMatchRowData = dataMap.get(hash(row));
+            if (mayMatchRowData == null) {
+                return false;
+            }
+            for (RowData mayMatch : mayMatchRowData) {
+                if (matchRow(row, mayMatch)) {
+                    return true;
+                }
+            }
+            return false;
+        }
+    }
+
+    private boolean matchRow(RowData row, RowData mayMatch) {
+        for (int i = 0; i < rowType.getFieldCount(); ++i) {
+            if (!Objects.equals(
+                    fieldGetters[i].getFieldOrNull(row),
+                    fieldGetters[i].getFieldOrNull(mayMatch))) {
```

Comment:
for Date, Time, Timestamp, the inner representation is int/long, but external representation is java.sql.Date, java.sql.Time, java.sql.Timestamp. We should make sure they can be compared. Please also add some tests in DynamicFilteringDataTest

Reply:
Values in the building rows and testing rows are all expected to be Flink internal types, which should be guaranteed by the builders and users. I suppose that's not the responsibility of the DynamicFilteringData. But maybe I can add some comments to explain this to notify all users.

Reply:
In addition, here users mean the users of this data class, rather than flink users. So I suppose the behavior of the users  should be under control.